File: frontend/src/app/models/quiz.ts
import 'reflect-metadata';
import { Solution } from './solution';
import { DataBase } from './database';

export class Attempt {
  id: number | undefined;
  timeStamp: Date | undefined;
  score: number | undefined;
  quizId: number | undefined;
  userId: number | undefined;
  answers: Answer[] = [];
}
export class Question {
    id: number | undefined;
    order: number | undefined;
    body!: string;
    quizId: number | undefined; // ID du quiz auquel appartient la question
    solutions: Solution[] = [];
    answers: Answer[] = [];
    isOpen: boolean = false;
}

export class Query{
  ColumnNames = [] = [];
}

export enum QuizStatus{
  EnCours = 0,
  Fini = 1,
  PasCommence = 2,
  Cloture = 3,
  Publié = 5,
  NonPublié = 6
}

export class Answer {
  id: number | undefined;
  sql?: string;
  timeStamp: Date | undefined;
  isCorrect: boolean = false;
  question: Question | undefined;
  attempt: Attempt | undefined;
}

export class Quiz {
    id: number | undefined;
    name?: string;
    description?: string;
    isPublished: boolean = false;
    isClosed: boolean = false;
    isTest: boolean = false;
    start?: Date;
    finish?: Date;
    database!: DataBase;
    databaseName?: string;
    databaseId: number | undefined;
    evaluation?: string;
    score?: string;
    status: QuizStatus = QuizStatus.PasCommence;
    questions: Question[] = [];
    attempts: Attempt[] = [];

    public get statusAsString(): string {
      return QuizStatus[this.status];
    }

    get display(): string {
      return `${this.name} - ${this.database} - ${this.start} - ${this.finish} - ${this.evaluation}- ${this.status}`;
  }
}

File: backend/Models/Quiz.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;
namespace prid_2324_a11.Models;

public enum QuizStatus{
    EnCours = 0,
    Fini = 1,
    PasCommence = 2,
    Cloture = 3,
    NonTest = 4,
    Publie = 5,
    NonPublie = 6,

}

public class Quiz{
    [Key]
    public int Id { get; set; }
    public string Name { get; set; } = "";
    public string? Description { get; set; } = "";
    public bool IsPublished { get; set; }
    public bool IsClosed { get; set; }
    public bool IsTest { get; set; }
    public DateTime? Start { get; set; }
    public DateTime? Finish { get; set; }

    [ForeignKey(nameof(DatabaseId))]
    public int DatabaseId { get; set; }
    public Database Database { get; set; } = null!;
    public virtual ICollection<Question> Questions { get; set; } = new HashSet<Question>();
    public virtual ICollection<Attempt> Attempts { get; set; } = new HashSet<Attempt>();

}
